houston/ardu/sandbox.py

```diff
@@ -102,59 +115,187 @@ def _start(self,
         # wait for longitude and latitude to match their expected values, and
         # for the system to match the expected `armable` state.
-        initial_lon = mission.initial_state['longitude']
-        initial_lat = mission.initial_state['latitude']
-        v = self.system.variable
+        initial_lon = self.state_initial['longitude']
+        initial_lat = self.state_initial['latitude']
+        initial_armable = self.state_initial['armable']
+        v = self.state_initial.__class__.variables
         while True:
-            observed = self.observe(0.0)
-            ready_lon = v('longitude').eq(initial_lon, observed['longitude'])
-            ready_lat = v('latitude').eq(initial_lat, observed['latitude'])
-            ready_armable = \
-                observed['armable'] == mission.initial_state['armable']
+            # self.observe()
+            ready_lon = v['longitude'].eq(initial_lon, self.state['longitude'])
+            ready_lat = v['latitude'].eq(initial_lat, self.state['latitude'])
+            ready_armable = self.state['armable'] == initial_armable
             if ready_lon and ready_lat and ready_armable:
                 break
             time.sleep(0.05)
 
-        if not self._post_connection_setup():
-            print("Post connection setup failed!")
+        if not self._on_connected():
+            raise Exception("post-connection setup failed.")  # FIXME better exception  # noqa: pycodestyle
 
         # wait until the vehicle is in GUIDED mode
         # TODO: add timeout
         guided_mode = dronekit.VehicleMode('GUIDED')
-        self.__connection.mode = guided_mode
-        while self.__connection.mode != guided_mode:
+        self.vehicle.mode = guided_mode
+        while self.vehicle.mode != guided_mode:
             time.sleep(0.05)
+        logger.debug("Mode done")
 
-    def _stop(self) -> None:
-        """
-        Closes any connection to a simulated vehicle inside the sandbox, before
-        terminating the associated simulator.
-        """
-        bzc = self.bugzoo.containers
-
+    def stop(self) -> None:
+        bzc = self._bugzoo.containers
         if self.connection:
             self.connection.close()
-
-        # close the SITL
         cmd = 'ps aux | grep -i sitl | awk {\'"\'"\'print $2\'"\'"\'} | xargs kill -2'  # noqa: pycodestyle
         bzc.command(self.container, cmd, stdout=False, stderr=False,
                     block=True)
 
-    def _post_connection_setup(self):
+    def _on_connected(self) -> bool:
         """
-        Instructions to run after the connection is established.
+        Called immediately after a connection to the vehicle is established.
         """
         return True
+
+    def run(self, commands: Sequence[Command]) -> 'MissionOutcome':
+        """
+        Executes a mission, represented as a sequence of commands, and
+        returns a description of the outcome.
+        """
+        from ..mission import MissionOutcome
```

Comment:
We shouldn't need to dynamically import `MissionOutcome`. It should be possible to lift this import statement to the top of the file?

Reply:
I remember the problem with this. `Mission` has dependency on `Sandbox` and `Sandbox` has dependency on `MissionOutcome`. But since `Mission` and `MissionOutcome` are in the same file, you cannot import `MissionOutcome` on top of `Sandbox`.

Reply:
You can't import `MissionOutcome` into `houston.sandbox`, but this is `houston.ardu.sandbox`, so you *should* be able to.